app/utils/kci_test/__init__.py

```diff
@@ -129,6 +132,19 @@ def save_or_update(doc, spec_map, collection, database, errors):
     return ret_val, doc_id
 
 
+def _add_test_log(dir_path, filename, log_data):
+    utils.LOG.info("Generating log files in {}".format(filename))
+    if not os.path.isdir(dir_path):
+        try:
+            os.makedirs(dir_path)
+        except OSError as e:
+            if e.errno != errno.EEXIST:
```

Comment:
Could also be handled with a positive test `if not os.path.exists(dir_path)` before calling `os.makedirs` rather than the `try-except` approach.

Reply:
I'm pretty much sure I used it for consistency with some other parts of the code, but surely it can be changed.

Reply:
Well if you expect the directory to not be there, then yes it can be useful to catch the exception.  But using try/except as part of the control flow is generally not great.  Maybe other parts of the code aren't very well written ;-)

Reply:
> But using try/except as part of the control flow is generally not great. 

Well, in general the answer is 'yes', but in Python specifically we have things like StopIteration or SystemExit exceptions which are used as a part of the control flow. Nonetheless, I'll change it to the conditional block.

Reply:
The whole code block looks like below. The exception there is used to silence all exceptions but the situation when the file exists and is not a directory. So the behavior is not the same as `if not os.path.exists(dir_path)`. The directory is not created when it exists, but raises an exception when there is a file which is named the same as the directory we want to create. This is consistent with the directory creation for lava tests.
```python
    if not os.path.isdir(dir_path):
        try:
            os.makedirs(dir_path)
        except OSError as e:
            if e.errno != errno.EEXIST:
                raise e
```

Reply:
With this code, it will call `os.makedirs` if there is a file called `dir_path` because it's not a directory, and then silence the exception because it can't create it.  I can see the same thing is being done in other places in the code but I can't make any sense of it, as I think it should be doing what you're saying.

```
$ touch hey && python -c "\
import os
import errno
  
dir_path = 'hey'

if not os.path.isdir(dir_path):
  try:
    os.makedirs(dir_path)
  except OSError, ex:
    print(errno.EEXIST, ex.errno)
    if ex.errno != errno.EEXIST:
      raise ex
"
(17, 17)
```

Surely if there is a file called `dir_path`, then writing the logs will fail just after that as you can't join file names together in `os.path.join()`.

Here's a patch that started doing this, but it doesn't explain why it was needed:
```diff
commit 0f3385a3589781bf05c1b961107b49683de8df4d
Author: Milo Casagrande <milo.casagrande@linaro.org>
Date:   Mon May 11 10:53:50 2015 +0200

    Handle corner case on directory creation.

diff --git a/app/utils/bootimport.py b/app/utils/bootimport.py
index c227b8e3bac9..6e33a4bb8717 100644
--- a/app/utils/bootimport.py
+++ b/app/utils/bootimport.py
@@ -23,6 +23,7 @@ except ImportError:
 import bson
 import copy
 import datetime
+import errno
 import os
 import re
 
@@ -144,7 +145,11 @@ def save_to_disk(boot_doc, json_obj, base_path):
 
     try:
         if not os.path.isdir(dir_path):
-            os.makedirs(dir_path)
+            try:
+                os.makedirs(dir_path)
+            except OSError, ex:
+                if ex.errno != errno.EEXIST:
+                    raise ex
 
         with open(file_path, mode="w") as write_json:
             write_json.write(
```

@MiloCasagrande Can you shed some light on this please?